### DISCUSSION TITLE
ESP32 switching I2S parallel to/from GPIO mode

### DISCUSSION THREAD

**Discussion by mcbp223** [lovyan03/LovyanGFX#320] [Q&A]
Hi, when trying to make the reading from display memory work, I wasn't able to until I've supplemented the mode switching functions beginRead and endRead with reconfiguration for each data pin, in addition to the RS signal, like

```cpp
for (size_t i = 0; i < 8; ++i)
{
    gpio_set_direction((gpio_num_t)pins[i], GPIO_MODE_INPUT);
    gpio_matrix_in(pins[i], 0x100, 0);
}
```
and
```cpp
for (size_t i = 0; i < 8; ++i)
{
  gpio_set_direction((gpio_num_t)pins[i], GPIO_MODE_OUTPUT);
  gpio_matrix_out(pins[i], idx_base + i, 0, 0);
}
```
respectively.

Has anyone encountered this, or it could be that my chip revision is old?

**Comment by tobozo** [comment]
gpio direction is [already set to both output and input](https://github.com/lovyan03/LovyanGFX/blob/develop/src/lgfx/v1/platforms/esp32/Bus_Parallel8.cpp#L135), does it also improve when using `GPIO_MODE_INPUT_OUTPUT` both in beginRead and endRead?

**Comment by mcbp223** [comment]
That's how I tried it first, with the library unchanged, data pins set to `GPIO_MODE_INPUT_OUTPUT` as you have indicated. With the  test_readpixel sketch I could see that there was something moving, but I don't know how to say it, it's kind of black on black, very faint, which really gave the impression of two outputs competing with each other on the same bus, so trying to set the MCU side explicitly to input only was the next step. If other ESP32 chips don't exhibit  this behavior then it could be something different with mine, maybe I'll try to wire a newer revision chip to this LCD and see if it still needs the bus input only or not, since the one I am using is maybe old, ESP32-D0WDQ6 Rev 1.

With the additions I've made, if in beginRead I replace `GPIO_MODE_INPUT` with `GPIO_MODE_INPUT_OUTPUT` it goes back to not working.

**Reply by mcbp223** [reply]
I've tried with an ESP32-D0WDQ5 Rev 3, but it still needs the data bus set to input only for reading.

Could this be due to my choice of wiring? I have `D0:D7=12:19`, `RST=25`, `CS=26`, `RD=27`, `RS=32`,  `WR=33`, with mcufriend style resistive touch connections at `Ym=D7=19`, `Yp=WR=33` and `Xp=D6=18`, `Xm=RS=32`.
